joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/passes/AstCreator.scala

```diff
@@ -2323,183 +2926,353 @@ class AstCreator(filename: String, javaParserAst: CompilationUnit, global: Globa
-  private def astForLambdaExpr(expr: LambdaExpr, order: Int): Ast = {
-    // TODO: Fix types in scope for lambdas
-    // TODO: Fix class name (currently `com.github.javaparser.ast.expr.LambdaExpr`)
-    val className = scopeStack.getEnclosingTypeDecl.map(_.fullName).getOrElse("<empty>")
-    val fullName  = s"$className:${nextLambdaName()}"
+  private def buildParamListForLambda(
+    expr: LambdaExpr,
+    maybeBoundMethod: Option[ResolvedMethodDeclaration],
+    expectedTypeParamTypes: Map[ResolvedTypeParameterDeclaration, ResolvedType]
+  ): Seq[Ast] = {
+    val lambdaParameters = expr.getParameters.asScala.toList
+    val paramTypesList = maybeBoundMethod match {
+      case Some(resolvedMethod) =>
+        val resolvedParameters = (0 until resolvedMethod.getNumberOfParams).map(resolvedMethod.getParam)
 
-    val parameterAsts       = astsForParameterList(expr.getParameters)
-    val namesToMethodParams = mapNamesToParams(parameterAsts)
+        // Substitute generic typeParam with the expected type if it can be found; leave unchanged otherwise.
+        val paramsWithSubstitutedTypes = resolvedParameters.map(_.getType).map {
+          case resolvedType: ResolvedTypeVariable =>
+            expectedTypeParamTypes.getOrElse(resolvedType.asTypeParameter(), resolvedType)
 
-    val bodyOrder = parameterAsts.size + 2
-    val bodyAst = if (expr.getBody.isBlockStmt) {
-      astsForStatement(expr.getBody, bodyOrder).headOption
-        .getOrElse(emptyBlock(bodyOrder))
-    } else {
-      val blockNode =
-        NewBlock()
-          .lineNumber(line(expr.getBody))
-          .columnNumber(column(expr.getBody))
-          .order(bodyOrder)
-          .argumentIndex(bodyOrder)
+          case resolvedType => resolvedType
+        }
 
-      val args = astsForStatement(expr.getBody, 1)
+        paramsWithSubstitutedTypes.map(typeInfoCalc.fullName)
 
-      Ast(blockNode).withChildren(args)
+      case None =>
+        // Unless types are explicitly specified in the lambda definition,
+        // this will yield the erased types which is why the actual lambda
+        // expression parameters are only used as a fallback.
+        lambdaParameters
+          .map(_.getType)
+          .map(typeInfoCalc.fullName)
+          .map(_.getOrElse(TypeConstants.UnresolvedType))
     }
 
-    val (identifiersMatchingParams, identifiersNotMatchingParams) = {
-      // bodyAst.ctx.identifiers.values.toSeq.partition(identifier => namesToMethodParams.contains(identifier.name))
-      (Nil, Nil)
+    if (paramTypesList.size != lambdaParameters.size) {
+      logger.error(s"Found different number lambda params and param types for $expr. Some parameters will be missing.")
     }
 
-    val closureBindings = closureBindingsForLambdas(identifiersNotMatchingParams)
-    val refEdgePairs    = buildRefEdgePairs(identifiersMatchingParams, namesToMethodParams)
+    val parameterNodes = lambdaParameters
+      .lazyZip(paramTypesList)
+      .lazyZip(LazyList.from(1))
```

Comment:
To answer this and the above comment: I did it this way, since the lazyZips give a result of `List[(NewParameter, String, Int)]` instead of `List[((NewParameter, String), Int)]`, the former being a bit cleaner in my opinion. I don't really have a strong opinion either way, however, so will update.